Mergin/clone_project_dialog.py

```diff
@@ -1,29 +1,66 @@
+from qgis.PyQt.QtCore import Qt
 from qgis.PyQt import uic
 
 ui_file = os.path.join(os.path.dirname(os.path.realpath(__file__)), "ui", "ui_clone_project.ui")
 
 
 class CloneProjectDialog(QDialog):
-    def __init__(self, username, user_organisations=None):
+    def __init__(self, user_info, workspaces=[], default_workspace=None):
         QDialog.__init__(self)
         self.ui = uic.loadUi(ui_file, self)
         self.ui.buttonBox.button(QDialogButtonBox.Ok).setEnabled(False)
         self.ui.buttonBox.accepted.connect(self.accept_dialog)
-        self.ui.projectNamespace.addItem(username)
-        if user_organisations:
-            self.ui.projectNamespace.addItems(
-                [o for o in user_organisations if user_organisations[o] in ["admin", "owner"]]
-            )
+
+        if workspaces is not None:
+            for ws in workspaces:
+                try:
+                    # Check if server is ee offering per workspace permissions
+                    is_writable = user_info["id"] in ws["writers"]
+                    self.ui.projectNamespace.addItem(ws["name"], is_writable)
+                except (KeyError, TypeError):
+                    # Server is ce, we'll ask for forgiveness instead of write permission
+                    is_writable = True
+                    self.ui.projectNamespace.addItem(ws, is_writable)
+        else:
+            # This means server is old and uses namespaces
+            username = user_info["username"]
+            user_organisations = user_info.get("organisations", [])
+            self.ui.projectNamespace.addItem(username, True)
+            self.ui.projectNamespaceLabel.setText("Owner")
+            for o in user_organisations:
+                if user_organisations[o] in ["admin", "owner"]:
+                    self.ui.projectNamespace.addItem(o, True)
+
+        self.ui.projectNamespace.currentTextChanged.connect(self.workspace_changed)
         self.ui.edit_project_name.textChanged.connect(self.text_changed)
 
+        # disable widgets if default workspace is read only
+        self.workspace_changed()
+        if default_workspace:
+            self.ui.projectNamespace.setCurrentText(default_workspace)
+
         # these are the variables used by the caller
         self.project_name = None
         self.project_namespace = None
 
     def text_changed(self):
-        self.ui.buttonBox.button(QDialogButtonBox.Ok).setEnabled(bool(self.ui.edit_project_name.text()))
+        enabled = bool(self.ui.edit_project_name.text()) and not self.ui.warningMessageLabel.isVisible()
+        self.ui.buttonBox.button(QDialogButtonBox.Ok).setEnabled(enabled)
+
+    def workspace_changed(self):
+        is_writable = bool(self.ui.projectNamespace.currentData(Qt.UserRole))
+        if is_writable:
+            msg = ""
+        else:
+            msg = "You do not have write permissions for this workspace"
```

Comment:
Not quite sure what means "writable". If user is a workspace writer, then he has permissions to push to all projects. If not, he can still write to projects which he was granted write permissions for. 

For creating new projects in workspace you need to be workspace admin. (However it used to be organization writer)

Reply:
OK, I got this wrong I guess. So let's get this straight.
In order to be able to clone into a workspace/namespace:
- in CE/EE the user id needs to be in *admins* or *owners*
- in old, the user needs to be in organization's *writers*?
The existing code prior to this PR was checking against organization's *admin* and *owner*
Can you confirm?

Reply:
- CE/EE - correct
- old: for destination namespace either it was user's namespace or user was organisation writer (checked latest server master)